### DISCUSSION TITLE
Support Array interface for Collections

### DISCUSSION THREAD

**Discussion by thiagomini** [mikro-orm/mikro-orm#4664] [General]
**Is your feature request related to a problem? Please describe.**
Yes, I can't treat my entity classes as "POJO" in unit tests. One of our goals while using Mikro-ORM is to avoid coupling entity behavior and logic with ORM accidental complexity - and that's why we are using schema definitions. However, when defining, for instance, that `User` has many `Books`, we can't rely on the Array definition:

```ts
public class User {
 books: Book[]
}
```

The issue really raised when we tried to use `books.find()` method, which the `Collection` wrapper does not implement. We could make use of the `Collection` class itself, but the problem is that it throws an error when trying to use it without initializing the MikroORM - so we can't have real unit tests, we depend on both the Class and the initialization of the ORM.

**Describe the solution you'd like**
I would like to be able to work with relationships as if they were arrays - so, the collection should implement the Array's interface. In our app, all relationships from the `One` side are eagerly loaded (because they are root of aggregates), and then we don't have to worry about these being loaded beforehand.

Alternatively, it would be even nicer if we could provide a custom `Collection` class that implements an interface. It might be useful to have our own custom methods to handle relationship between entities.

**Describe alternatives you've considered**
* Tried to use the `Collection` class, but unit tests will fail because we haven't initialized the ORM.

**Additional context**
Keep in mind that the premise we're trying to achieve here is to treat every entity and relationship as if they were in memory. That's why it's important to rely on the default Array interface, otherwise, our domain logic will be coupled to accidental details.

**Comment by B4nan** [comment]
> Tried to use the Collection class, but unit tests will fail because we haven't initialized the ORM.

Why is that a problem to initialize the ORM? You know you can just initialize the ORM without connecting to the database? In v6 there is also a sync variant of the `init` method which initializes the connection lazily when needed. 

I guess we could remove the validation when you work with the `Collection` and just not propagate when the metadata are not available, I also don't like the fact that the ORM init is required to work with the classes, it indeed feels weird. At the same time, I don't like that the behavior would change, so maybe it would be better to have an opt-out of this validation.

The `Collection` class does a lot of things under the hood, including change tracking and propagation of changes to the other side of a bidirectional relation, I don't think we can just map to arrays (at least not without a huge refactor that I honestly don't want to undergo). But I don't mind adding more methods to the `Collection` class, in fact, the next version will have a few more helpers in there, like `map/reduce/filter`, it was always meant to have an array-like interface (e.g. it has array index access). We can also add shortcuts like `push` -> `add`, but I would rather stick to the reading methods instead of mutation.

**Reply by thiagomini** [reply]
Hey, Martin, I appreciate the reply! It's not exactly a problem to initialize the ORM, but it's just an inconvenience, given that the unit tests should be isolated from it. And opting out for metadata validation when working with collections would also require configuring the ORM for unit tests 🤔 

I guess that we could have a way of defining our own implementation of collection. I mean, the ORM users might not want to depend on this specific implementation of collections and work with something simpler. I believe we could do it by providing an interface of `Collection` that can work with both database and in-memory-only storage.

In any case, expanding the interface of the Collection might indeed be a good feature!

**Reply by B4nan** [reply]
> And opting out for metadata validation when working with collections would also require configuring the ORM for unit tests 🤔

Could be just an env var.

Technically the hydration is already extensible, you can bring your own implementation, and there you could alter how the entities are mapped. But change tracking won't work if you don't use the `Collection` class...

**Reply by B4nan** [reply]
Actually, among the new ones, there is already `findFirst`, I guess it's better to call this just `find` so it's aligned with the `Array` interface.

https://github.com/mikro-orm/mikro-orm/commit/f0ede2d95c8609167d14a1ccf1a7f7502a89a626

**Reply by thiagomini** [reply]
Hmmmm I see, could an env var work to make the metadata not to be validated when the ORM is not initialized then? Just so we can work with it in unit tests 🤔

**Reply by thiagomini** [reply]
Hey again, @B4nan . You mentioned there's a way to start Mikro ORM without connecting to the database. How can I do that? 🤔

**Reply by drackp2m** [reply]
How interesting that you started talking about this topic just three days ago.

> You know you can just initialize the ORM without connecting to the database?

No @B4nan, I didn't know that, and I'm really glad because it will surely get me out of the fix I'm in.

I have a very similar problem to the one @thiagomini just mentioned. I have a Game entity with a ManyToMany relationship with another User entity, and when running my unit tests, I found that I can't call the add method of the collection:

`Cannot read properties of undefined (reading '__meta')`

If I simply import `MikroOrmModule.forRoot()` in the initialization of the TestingModule of NestJS, the tests pass, but the execution time suffers considerably, also in the pipelines of Github Actions, where until now I did not require DB configuration, I get an error for not having loaded the environment variables.

I have been looking in the documentation, on the internet and playing with the `MikroORM.init` method, but I have not achieved any favorable result, so I would appreciate any clarification from your side.

**Reply by thiagomini** [reply]
Hey @drackp2m , I actually found a solution. There's an option called `connect` for `MikroORMModule`. Just set it to `true` and you should be good to go. Here's my snippet for a utility function that should be called in the `beforeAll` for our tests:

```ts
export async function registerMikroOrmMetadata() {
  const shouldConnectToDatabase = false;
  const orm = await MikroORM.init(
    {
      type: 'postgresql',
      clientUrl: 'postgresql://postgres:postgres@localhost:5432/postgres', // that string does not matter - we are not connecting to the database during unit tests
      allowGlobalContext: true,
      discovery: {
        warnWhenNoEntities: false,
        checkDuplicateEntities: false,
      },
      entities: schemas,
    },
    shouldConnectToDatabase,
  );
  registerMikroFactories(orm);
  return orm;
}
```

⚠️ Notice that here I'm initializing the `MikroORM` directly, without calling the `MikroORMModule`. If you are not using any Nest feature for your unit tests, you shouldn't need the `MikroORMModule` to be initialized.

**Reply by thiagomini** [reply]
By the way, the `registerMikroFactories` step is just calling my test data generator library [factory-girl-ts](https://github.com/thiagomini/factory-girl-ts), if you want to give it a look :)

**Reply by drackp2m** [reply]
Good @thiagomini, thank you very much for sharing your code, it has helped me. I would never have imagined that MikroORM.init would have a second parameter.

However, there is still something I am not quite happy with, what does your `schemas` variable contain?

I say this because, it was enough for me to place this just in front of the initialization of the NestJS test module:

```typescript
await MikroORM.init(
  defineConfig({
    clientUrl: 'postgresql://user:pass@localhost/db_name',
    entities: [Game],
  }),
  false,
);
```

If I leave entities with a path where it will not find anything, and I add `discovery: { warnWhenNoEntities: false },` to the configuration, I still get the same message as before.


The question is, if I don't pass to the array of entities, at least one valid entity, I get the error, however, this is not something totally tragic, much better than before of course!!!!

**Reply by B4nan** [reply]
FYI the second parameter is already removed from v6 in favor of the `connect` option - which is already part of v5 for quite some time.

https://mikro-orm.io/api/core/interface/MikroORMOptions#connect

**Reply by drackp2m** [reply]
Oh, as I hardly ever go in to view the documentation from the API, I didn't notice, thanks for the link, I'll change it as soon as I get home.

Anyway, I understand that it must be very time consuming to keep all the documentation up to date and synchronized, but....
- In case the second parameter of `MikroORM.init()` is going to be removed in v6, shouldn't it already appear as `deprecated` with the small explanation of what should be used instead?
- In [this section](https://mikro-orm.io/docs/configuration#connection) shouldn't it also be explained?
- In the API documentation you have passed me, shouldn't the connect parameter appear as optional, and maybe also, its default value, which I imagine will be `true`?
